src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -125,57 +72,90 @@ new fs mergeLevel mergeMappend targetPaths runs = do
 
 data StepResult fhandle = MergeInProgress | MergeComplete !(Run fhandle)
 
--- | Do a fixed number of steps of merging. Each step reads a single entry, then
--- either resolves it with the previous one or writes the previous one out to
--- the run being created.
+stepsInvariant :: Int -> (Int, StepResult a) -> Bool
+stepsInvariant requestedSteps = \case
+    (n, MergeInProgress) -> n >= requestedSteps
+    _                    -> True
+
+-- | Do at least a given number of steps of merging. Each step reads a single
+-- entry, then either resolves the previous entry with the new one or writes it
+-- out to the run being created.
+-- Returns the number of input entries read, which is guaranteed to be at least
+-- as many as requested (unless the merge is complete).
```

Comment:
```suggestion
-- | Do at least a given number of steps of merging. Each step reads exactly one
-- entry, and outputs zero or one entries to the run under construction. 
-- Returns the number of input entries read, which is guaranteed to be at least
-- as many as requested (unless the merge is complete).
```

And maybe we should be even more explicit about why it might do more steps than requested